### ISSUE TITLE
How to use a different metric for the EarlyStopping?

### ISSUE THREAD

**Issue by coder-rakeshpaul** [ultralytics/ultralytics#5925]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hello,

according to answer in issue #4521 I understand that the validation loss is the criteria for EarlyStopping

Is there a way how we could change the metric that is used as the criteria for EarlyStopping?? While I understand the reasoning behind using the validation loss as the metric. I would like to change it and use a different metric for the criteria

**Comment by glenn-jocher** [comment]
@coder-rakeshpaul hello,

To change the criteria for EarlyStopping in YOLOv8, you must modify the code in the training process. Currently, the criteria is set to check the validation loss for early stopping, as it often correlates to good performance in prediction tasks.

To change this, you'd have to locate the part of the code where EarlyStopping is defined, and adjust it to use a metric of your preference. The choice of metric to use would depend on your specific needs and objectives.

Metrics that could potentially be used as criteria in early stopping include precision, recall, f1-score, or other custom evaluation metrics suitable for your project. Please be aware that the use of metrics other than validation loss may not necessarily improve the final performance of the model.

Hope this provides sufficient guidance on modifying the EarlyStopping criteria. Be reminded that altering core functionality like this should be done carefully, with adequate testing to ensure that it behaves as expected.

Best of luck with your project!

**Comment by coder-rakeshpaul** [comment]
Thanks. could you tell the files that I might have to look into in in order to modify this metric?

Inside the [torch_utils.py](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/utils/torch_utils.py) in the class EarlyStopping I found this block of code and this seems to be monitoring the mAP as the criteria
 
`class EarlyStopping:
    """Early stopping class that stops training when a specified number of epochs have passed without improvement."""

    def __init__(self, patience=50):
        """
        Initialize early stopping object.

        Args:
            patience (int, optional): Number of epochs to wait after fitness stops improving before stopping.
        """
        self.best_fitness = 0.0  # i.e. mAP
        self.best_epoch = 0
        self.patience = patience or float('inf')  # epochs to wait after fitness stops improving to stop
        self.possible_stop = False  # possible stop may occur next epoch

    def __call__(self, epoch, fitness):
        """
        Check whether to stop training.

        Args:
            epoch (int): Current epoch of training
            fitness (float): Fitness value of current epoch

        Returns:
            (bool): True if training should stop, False otherwise
        """
        if fitness is None:  # check if fitness=None (happens when val=False)
            return False

        if fitness >= self.best_fitness:  # >= 0 to allow for early zero-fitness stage of training
            self.best_epoch = epoch
            self.best_fitness = fitness
        delta = epoch - self.best_epoch  # epochs without improvement
        self.possible_stop = delta >= (self.patience - 1)  # possible stop may occur next epoch
        stop = delta >= self.patience  # stop training if patience exceeded
        if stop:
            LOGGER.info(f'Stopping training early as no improvement observed in last {self.patience} epochs. '
                        f'Best results observed at epoch {self.best_epoch}, best model saved as best.pt.\n'
                        f'To update EarlyStopping(patience={self.patience}) pass a new patience value, '
                        f'i.e. `patience=300` or use `patience=0` to disable EarlyStopping.')
        return stop`

Is this not the file I should be looking at and some other??

**Comment by glenn-jocher** [comment]
@coder-rakeshpaul, you're on the right track. The class `EarlyStopping` within the `torch_utils.py` file is indeed responsible for the early stopping mechanism. As you noticed, this class checks the improvement between epochs based on a 'fitness' metric which is currently set to mAP (mean Average Precision).

To modify the metric for early stopping, you need to pass a different metric value to the __call__ method of the `EarlyStopping` class. When you train your model, the value of the fitness (the metric you choose) at each epoch is evaluated inside this method.

When you modify your code, make sure to update the actual calculation of fitness value in accordance with the metric you decide on. Also, please remember to update the rest of the __call__ method and __init__ method accordingly, if necessary.

Keep in mind that changing the early stopping criteria may not necessarily lead to better model performance, so thorough testing and evaluation is recommended. Good luck with the modifications!

**Comment by coder-rakeshpaul** [comment]
@glenn-jocher thank you for your guidance.

There is another question previously you mentioned that the criteria was Validation Loss, was that some sort of error or is it something which I have not fully understood?? As I can't seem to find anything related to validation loss here.

**Comment by glenn-jocher** [comment]
Hello @coder-rakeshpaul,

Thank you for your question! Indeed, it seems there might have been some confusion in our previous communication. In YOLOv8, the early stopping criteria is evaluated using a fitness metric, which is currently set to the Mean Average Precision (mAP), not the validation loss. I apologize for any confusion caused by this.

The mAP is a common metric used in object detection tasks to quantify the performance of the model across all classes. Using mAP as the early stopping criteria means that the training will stop if the model's mAP performance does not improve for a specific number of epochs (patience).

Unfortunately, there is no usage of validation loss for the early stopping process in the current YOLOv8 implementation. If you wish to modify this, you would need to change the fitness metric to validation loss in the EarlyStopping class.

I hope this clarifies your concern. Please feel free to ask if you have any more questions. Happy coding!

**Comment by coder-rakeshpaul** [comment]
Hey @glenn-jocher,

Thanks for the clarification.

**Comment by glenn-jocher** [comment]
Hello @coder-rakeshpaul,

You're welcome! I'm glad I could help clarify things for you. If you have any more questions or need further assistance while working with YOLOv8, don't hesitate to ask. Happy coding and all the best with your project!
